### PR TITLE
sql: set survive zone on regional by row system tables

### LINKED ISSUE #113296: bootstrap: some system tables don't inherit constraints/preferences from system database
The following tables don't inherit constraints and lease preferences from the system database zone config by default:

* `replication_constraint_stats `
* `replication_stats`
* `tenant_usage`

This is inconvenient since users must explicitly set constraints and lease preferences for these specific tables in addition to the system zone.

This is likely because these default zone configs are not created via `NewZoneConfig` (which enables inheritance by default), nor derived from the default system zone configs, but instantiated directly with inheritance defaulting to `false`:

https://github.com/cockroachdb/cockroach/blob/a87f0c71ab05a1569641744cb687e8c913d46f1b/pkg/sql/catalog/bootstrap/metadata.go#L541-L550

The tell-tale sign is that `SHOW ZONE CONFIGURATIONS` will show explicit empty constraints/preferences for these tables, but not for other system tables -- compare e.g.:

```
ALTER TABLE system.public.replication_stats CONFIGURE ZONE USING
    gc.ttlseconds = 600,
    constraints = '[]',
    lease_preferences = '[]';


ALTER TABLE system.public.statement_statistics CONFIGURE ZONE USING
    gc.ttlseconds = 3600;
```

We'll likely need to both fix the initial creation to inherit, and also add a migration to fix these in existing clusters.

---

Addendum 2024-02-16: It would also be nice while we are cleaning all this up to add a migration to lower the gc.ttl of `system.lease` as well. See [this thread](https://cockroachlabs.slack.com/archives/C04N0AS14CT/p1708104269274429?thread_ts=1708103486.280399&cid=C04N0AS14CT).

---

Jira issue: CRDB-32857

Epic CRDB-33032

### LINKED ISSUE #119090: sql: modify system database tables to set survival goals
Now that region liveness support has been added, we can finally adopt SURVIVE ZONE FAILURE for the system.lease, system.sqlliveness, system.sqlinstances tables. Additionally, we can remove temporarily logic that demotes the system database from: https://github.com/cockroachdb/cockroach/pull/109844

This change should also add a migration to set these goals on existing tables.

Jira issue: CRDB-36031

Epic CC-24173

### PR DESCRIPTION
**Pull request by fqazi**
Now that the region liveness work has been work we can finally enable it on multi-region clusters. To help aid in this this patch will do the following:

1.  All system tables will now be assigned a locality out of the gate, previously we were missing logic to set these settings on newly added system tables.
2. Adding a region on the system database will configure the system database as SURVIVE REGION. Also the table: system.lease, system.sqlliveness, system.sqlinstances will be configured as survive zone.
3. An upgrade will be added to migrate existing system database to SURVIVE REGION and the same version gate will be used to block setting SURVIVE REGION incorrectly. The tables will setup as system.lease, system.sqlliveness, system.sqlinstances will be migrated to SURVIVE ZONE.
4. Previously, some system tables incorrectly did not have the inherit flags set on their zone configurations, these will be fixed during migration and during bootstrap.g
5. The system.lease session based upgrade had a bug which caused it incorrectly be setup on multi-region configurations by not propagating partition information on the primary index.

Fixes: #119090
Fixes: #113296
Epic: CC-24173

### CODE REVIEW

**Review by rafiss — CHANGES_REQUESTED** [review]
this is fitting together pretty nicely!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-:-NqcGdRo9OlucnAS7GnU:b-ndw347)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @JeffSwenson)

___
*[`pkg/sql/alter_database.go` line 1215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcEiaoBou_64hBAYeN:-NqcEiaoBou_64hBAYeO:b-ujtpfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc42b5f3c95078579c0b67a84afea0a1364ffaed/pkg/sql/alter_database.go#L1215)):*
> ```Go
> 		return nil
> 	}
> 	if !params.p.EvalContext().Settings.Version.IsActive(params.ctx, clusterversion.V24_1_SystemDatabaseSurvivability) {
> ```

perhaps a comment here or on the `maybeUpdateSystemDBSurvivalGoal` function could help, saying why we don't want to call it anymore. (and similar for the other places where `maybeUpdateSystemDBSurvivalGoal` is called)
___
*[`pkg/sql/region_util.go` line 2490 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcCwzY6iN_5JzfLeKs:-NqcCwzY6iN_5JzfLeKt:b-xdxv77) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6b7ddb6f55a2886b485172e641b0dfad4a18216/pkg/sql/region_util.go#L2490)):*
> ```Go
> 			continue
> 		}
> 		globalTables = append(globalTables, name)
> ```

hm do we really want everything else to be global? i'm concerned a little about some of the write-heavy tables

why don't we want any to just be regional tables?
___
*[`pkg/upgrade/upgrades/upgrades.go` line 181 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcFP3ND6PL_6BX_1dN:-NqcFP3O5_Q9DXDG3nJJ:bo0g33k) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc42b5f3c95078579c0b67a84afea0a1364ffaed/pkg/upgrade/upgrades/upgrades.go#L181)):*
> ```Go
> 	),
> 
> 	upgrade.NewTenantUpgrade(
> ```

i don't understand when we should use `upgrade.NewTenantUpgrade` vs `upgrade.NewSystemUpgrade`. do you know why tenant upgrade is correct here?
___
*[`pkg/upgrade/upgrades/upgrades.go` line 186 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcFi4zCH7SJ-iZxRW6:-NqcFi4zCH7SJ-iZxRW7:blvctap) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc42b5f3c95078579c0b67a84afea0a1364ffaed/pkg/upgrade/upgrades/upgrades.go#L186)):*
> ```Go
> 		upgrade.NoPrecondition,
> 		alterSystemDatabaseSurvivalGoal,
> 		upgrade.RestoreActionNotRequired("cluster restore does not restore the system database"),
> ```

technically, it does restore some things in the system database. should this say "cluster restore does not preserve the multiregion configuration of the system database" ?
___
*[`pkg/upgrade/upgrades/v23_2_create_region_liveness.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcDSyqC126kP0UW0kY:-NqcDSyqC126kP0UW0kZ:b19933e) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6b7ddb6f55a2886b485172e641b0dfad4a18216/pkg/upgrade/upgrades/v23_2_create_region_liveness.go#L54)):*
> ```Go
> 	}
> 	if err := createSystemTable(ctx, d.DB, d.Settings, d.Codec,
> 		regionLivenessMutable, false /*setGlobalLocality*/); err != nil {
> ```

since this is false, should that mean that `region_liveness` should appear in the `rbrTables` list in `optimizeSystemDatabase`?
___
*[`pkg/upgrade/upgrades/v24_1_system_database.go` line 33 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcFxRfEp8Afb77APv3:-NqcFxRfEp8Afb77APv4:bpo4r39) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc42b5f3c95078579c0b67a84afea0a1364ffaed/pkg/upgrade/upgrades/v24_1_system_database.go#L33)):*
> ```Go
> 			return err
> 		}
> 		if !systemDB.IsMultiRegion() {
> ```

would it be easier to check this by using a query like `show regions from database system;`? i'm fine with it as is too; since accessing the collection directly might be faster?
___
*[`pkg/ccl/multiregionccl/multiregion_system_table_test.go` line 506 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcGVjs8-AysL3xFscS:-NqcGVjs8-AysL3xFscT:bb9vcxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc42b5f3c95078579c0b67a84afea0a1364ffaed/pkg/ccl/multiregionccl/multiregion_system_table_test.go#L506)):*
> ```Go
> 	_, err = cluster.Conns[0].Exec("SET CLUSTER SETTING version = crdb_internal.node_executable_version();")
> 	require.NoError(t, err)
> 	tDB.Exec(t, "SET CLUSTER SETTING version = crdb_internal.node_executable_version();")
> ```

nit: do we need something here to keep polling `SHOW CLUSTER SETTING version` until the upgrade is complete?
___
*[`pkg/upgrade/upgrades/descriptor_utils_test.go` line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcE7IGBlL_j4VvF9N8:-NqcE7IGBlL_j4VvF9N9:bxfwqss) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6b7ddb6f55a2886b485172e641b0dfad4a18216/pkg/upgrade/upgrades/descriptor_utils_test.go#L96)):*
> ```Go
> 	descDB := tc.Server(0).InternalDB().(descs.DB)
> 	require.NoError(t, upgrades.CreateSystemTable(
> 		ctx, descDB, tc.Server(0).ClusterSettings(), keys.SystemSQLCodec, table, true, /*setLocality*/
> ```

nit: the comment should say `setGlobalLocality`, so it's more clear what the bool argument does
___
*[`pkg/upgrade/upgrades/descriptor_utils_test.go` line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcEBOVBLAi-Sq7duNm:-NqcEBOVBLAi-Sq7duNn:bxfwqss) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6b7ddb6f55a2886b485172e641b0dfad4a18216/pkg/upgrade/upgrades/descriptor_utils_test.go#L105)):*
> ```Go
> 	// Make sure it's idempotent.
> 	require.NoError(t, upgrades.CreateSystemTable(
> 		ctx, descDB, tc.Server(0).ClusterSettings(), keys.SystemSQLCodec, table, true, /*setLocality*/
> ```

nit: the comment should say `setGlobalLocality`, so it's more clear what the bool argument does

**Comment by rafiss** [comment]
this is totally unrelated to your PR, but while we are adding one system database upgrade migration, do you think we could fix https://github.com/cockroachdb/cockroach/issues/113296 as well?

**Comment by fqazi** [comment]
> this is totally unrelated to your PR, but while we are adding one system database upgrade migration, do you think we could fix #113296 as well?

Yeah I'll sort that out here as well

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119182)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson and @rafiss)

___
*[`pkg/sql/alter_database.go` line 1215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcEiaoBou_64hBAYeN:-NqjxBQr0FvoY1ETC4LI:b-48mu87) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc42b5f3c95078579c0b67a84afea0a1364ffaed/pkg/sql/alter_database.go#L1215)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

perhaps a comment here or on the `maybeUpdateSystemDBSurvivalGoal` function could help, saying why we don't want to call it anymore. (and similar for the other places where `maybeUpdateSystemDBSurvivalGoal` is called)
</blockquote></details>

Added the comment:

// Now that the zone survival goal is properly supported on the system database,  
// with region liveness support we no longer to inherit the stronger guarantees  
// assigned to other databases. i.e. Previously if any database had survive region,  
// the system database would be forced to inherit those stronger guarantees.
___
*[`pkg/sql/region_util.go` line 2490 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcCwzY6iN_5JzfLeKs:-Nr3gLpU8THjqTCy0ZyF:b45b6b3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6b7ddb6f55a2886b485172e641b0dfad4a18216/pkg/sql/region_util.go#L2490)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm do we really want everything else to be global? i'm concerned a little about some of the write-heavy tables

why don't we want any to just be regional tables?
</blockquote></details>

Done.

Good point, I made the other tables as by region.
___
*[`pkg/upgrade/upgrades/upgrades.go` line 181 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcFP3ND6PL_6BX_1dN:-Nr3gYQAA6VRNcvfM4tc:b-dmfs6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc42b5f3c95078579c0b67a84afea0a1364ffaed/pkg/upgrade/upgrades/upgrades.go#L181)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i don't understand when we should use `upgrade.NewTenantUpgrade` vs `upgrade.NewSystemUpgrade`. do you know why tenant upgrade is correct here?
</blockquote></details>

The system database exists in both system tenants and other tenants, so thats why need to be opt for this one.
___
*[`pkg/upgrade/upgrades/upgrades.go` line 186 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcFi4zCH7SJ-iZxRW6:-Nr3gqs6AJwuhLXzairD:b-k3p230) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc42b5f3c95078579c0b67a84afea0a1364ffaed/pkg/upgrade/upgrades/upgrades.go#L186)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

technically, it does restore some things in the system database. should this say "cluster restore does not preserve the multiregion configuration of the system database" ?
</blockquote></details>

Done.

Good point
___
*[`pkg/upgrade/upgrades/v23_2_create_region_liveness.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcDSyqC126kP0UW0kY:-Nr3hQ5b3cp5TwrWIl6M:b-kexqqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6b7ddb6f55a2886b485172e641b0dfad4a18216/pkg/upgrade/upgrades/v23_2_create_region_liveness.go#L54)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

since this is false, should that mean that `region_liveness` should appear in the `rbrTables` list in `optimizeSystemDatabase`?
</blockquote></details>

Refactored the logic here, but this system table is intentionally global.
___
*[`pkg/upgrade/upgrades/v24_1_system_database.go` line 33 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcFxRfEp8Afb77APv3:-Nr3l0kYDiRz6e2xlydz:b-o82psm) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc42b5f3c95078579c0b67a84afea0a1364ffaed/pkg/upgrade/upgrades/v24_1_system_database.go#L33)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

would it be easier to check this by using a query like `show regions from database system;`? i'm fine with it as is too; since accessing the collection directly might be faster?
</blockquote></details>

I think the two approaches are equivalent, this one probably will fetch fewer descriptors since its not layered on top of:  crdb_internal.databases.
___
*[`pkg/ccl/multiregionccl/multiregion_system_table_test.go` line 506 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcGVjs8-AysL3xFscS:-Nr3myvd0kGDB0DsU4PQ:b-h39q6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc42b5f3c95078579c0b67a84afea0a1364ffaed/pkg/ccl/multiregionccl/multiregion_system_table_test.go#L506)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: do we need something here to keep polling `SHOW CLUSTER SETTING version` until the upgrade is complete?
</blockquote></details>

The "SET CLUSTER SETTING version=..." is a blocking operation and reports any issues synchronously, so that shouldn't be needed here.
___
*[`pkg/upgrade/upgrades/descriptor_utils_test.go` line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NqcE7IGBlL_j4VvF9N8:-Nr3hfiS5x542-jNZMql:b884kni) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6b7ddb6f55a2886b485172e641b0dfad4a18216/pkg/upgrade/upgrades/descriptor_utils_test.go#L96)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: the comment should say `setGlobalLocality`, so it's more clear what the bool argument does
</blockquote></details>

Done.

Replaced the flag with a tree.LocalityLevel

**Comment by fqazi** [comment]
@rafiss  Reworked things a bit, since after chatting with Jeff we need to leave the system database as SURVIVE REGION, but only configure SURVIVE ZONE on the regional by row tables.

@JeffSwenson RFAL

File: pkg/upgrade/upgrades/v24_1_system_database.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/upgrade"
+)
+
+// alterSystemDatabaseSurvivalGoal sets the survival goal on the system database
+// to be SURVIVE ZONE.
```
Comment (JeffSwenson) [inline pkg/upgrade/upgrades/v24_1_system_database.go:26]: nit: this comment should read SURVIVE REGION

File: pkg/sql/region_util.go [hunk 2]
```diff
@@ -2505,6 +2579,10 @@ func (p *planner) optimizeSystemDatabase(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
+	/*
```
Comment (JeffSwenson) [inline pkg/sql/region_util.go:2582]: nit: delete the commented out code.

**Review by jeffswenson — APPROVED** [review]
LGTM

🥳 it's exciting to see the end of the project. Once this is merged I'll spin up a serverless host to perform some manual e2e testing.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119182)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @fqazi, @mgartner, and @msbutler)

___
*[`pkg/sql/region_util.go` line 2539 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-Nrph2QFAEublgeBGOim:-Nrph2QFAEublgeBGOin:btbxpqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3d90765467926ede3c378cc38e7fd7cf0ff178b/pkg/sql/region_util.go#L2539)):*
> ```Go
> 		"settings":               {},
> 		"descriptor":             {},
> 		"namespace":              {},
> ```

just double-confirming: is there a risk to making descriptor and namespace global? that will make all schema changes much slower, right? with descriptor leasing, is it possible that the added benefit of a global descriptor table is less important, since most of the reads of descriptors will be cached? (i don't have an opinion yet.. trying to figure out what's best)

same goes for role_members making all GRANTs slower, but that seems more ok
___
*[`pkg/sql/catalog/bootstrap/metadata.go` line 543 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NrpieOTA1jFCkTFvSka:-NrpieOTA1jFCkTFvSkb:bj562r5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0893e0e1ad6c5ab12315a67add64fe8a8cfc450e/pkg/sql/catalog/bootstrap/metadata.go#L543)):*
> ```Go
> 
> 	tenantUsageZoneConf := zonepb.NewZoneConfig()
> 	tenantUsageZoneConf.GC = &zonepb.GCPolicy{TTLSeconds: int32(systemschema.TenantUsageTableTTL.Seconds())}
> ```

while we're fixing these, should we also update system.lease to have a smaller gc.ttl?

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119182)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mgartner, @msbutler, and @rafiss)

___
*[`pkg/sql/region_util.go` line 2539 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-Nrph2QFAEublgeBGOim:-NrpsfT79SE1X16TIork:b-8peynu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3d90765467926ede3c378cc38e7fd7cf0ff178b/pkg/sql/region_util.go#L2539)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

just double-confirming: is there a risk to making descriptor and namespace global? that will make all schema changes much slower, right? with descriptor leasing, is it possible that the added benefit of a global descriptor table is less important, since most of the reads of descriptors will be cached? (i don't have an opinion yet.. trying to figure out what's best)

same goes for role_members making all GRANTs slower, but that seems more ok
</blockquote></details>

It will make descriptor creation / updates slightly slower because those will involve global tables. The main benefit will be to anything involving pg_catalog / information_schema, since we don't have to do cross node hops. The regional by row lease table will probably help making things a bit more snappier and make other parts slightly better.

So overall I think query workloads will perform better if they have introspection. Schema changes may be slightly slower, but these weren't fast operations already and have to wait for jobs to startup and leases to expire.
___
*[`pkg/sql/catalog/bootstrap/metadata.go` line 543 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-NrpieOTA1jFCkTFvSka:-NrptRwjA0nEEtDy0Sqf:b12a5bj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0893e0e1ad6c5ab12315a67add64fe8a8cfc450e/pkg/sql/catalog/bootstrap/metadata.go#L543)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

while we're fixing these, should we also update system.lease to have a smaller gc.ttl?
</blockquote></details>

Good point, let me add that as well.

**Review by rafiss — APPROVED** [review]
really nice work!

> Reviewed 6 of 20 files at r4, 13 of 15 files at r17, 10 of 11 files at r18, 7 of 7 files at r20, 1 of 1 files at r21, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119182#-:-NrqBVPf8JPGUyM4Qz-7:b-tqdyb)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @mgartner, and @msbutler)

**Comment by fqazi** [comment]
@rafiss @JeffSwenson TFTR!

**Comment by RaduBerinde** [comment]
We forgot to bump `SystemDatabaseSchemaBootstrapVersion` here.. Note that this was mentioned in `TestSystemDatabaseSchemaBootstrapVersionBumped` which was modified.

**Pull request merged**